legacy/reqcounter/reqcounter.go

```diff
@@ -16,30 +16,73 @@ limitations under the License.
 
 package reqcounter
 
-import "sync"
+import (
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
 
 // RequestCounter records the number of HTTP requests to GCR.
 type RequestCounter struct {
-	mutex    sync.Mutex
-	requests uint64
+	mutex    sync.Mutex // Lock to prevent race-conditions with concurrent processes.
+	requests uint64     // Number of HTTP requests since recording started.
+	quota    uint64     // Maximum requests allowed before GCR throttling occurs.
+	since    time.Time  // When the current counter began recording requests.
+}
+
+// log records the number of HTTP requests found and resets the counter.
+func (rc *RequestCounter) log() {
+	// Grab the lock to prevent race-conditions.
+	rc.mutex.Lock()
+	defer rc.mutex.Unlock()
+	// Log the number of requests found.
+	msg := fmt.Sprintf("Since %s there have been %d requests to GCR.", counter.since.Format("01-02-2006 15:04:05"), counter.requests)
+	logrus.Debug(msg)
+	// Reset the counter.
+	rc.reset()
+}
+
+// reset clears the number of requests and stamps the object with the current time of reset.
+// WARNING: To avoid race conditions, only use this method when holding the counter's mutex.
+func (rc *RequestCounter) reset() {
+	rc.requests = 0
+	rc.since = time.Now()
 }
 
+const (
+	// measurementWindow specifies the length of time to batch requests to GCR. Since Google's
+	// Container Registry specifies a quota of 50,000 HTTP requests per 10 min, the window
+	// for recording requests is set to 10 min.
+	// Source: https://cloud.google.com/container-registry/quotas
+	measurementWindow = time.Minute * 10
```

Comment:
1. How do we know that the 10 minutes that we measure is the the same 10 minute block measured by GCP? Does this matter?

Trying to figure out this question myself, I looked at https://cloud.google.com/container-registry/quotas and then clicked on the link to https://cloud.google.com/docs/quota. There is a lot of info there and it appears that quotas are already automatically monitored by GCP (see https://cloud.google.com/docs/quota#viewing_quota).

BUT according to this heading https://cloud.google.com/docs/quota#checking_metric_support it's unclear if GCR quotas (which seem to be tied to Google Cloud Storage quotas) are automatically monitored. FWIW I have seen the `429 TOO MANY REQUESTS` error (see https://cloud.google.com/docs/quota#quota_errors) before from the auditor, so we have definitely hit the quota limit before.

2. Can you just double-check that HTTP requests to GCR are *not* visible even with the `Quota` tool in GCP, under the top-level `IAM & Admin` tool in GCP?

Reply:
1. I am not going to assume the 10min window we are logging at will line up with GCP's quota. This feature is merely a benchmark to aid in the understanding of how many times the auditor is requesting GCP. Therefore, it is possible for our logs to reflect just under the quota limit, when in actuality we surpassed the quota and are being throttled. But this is okay, as these logs will at least tell us that we got VERY close to the quota limit during that time period.

Reply:
2. After offline discussion, we both discovered we can't view these HTTP requests to GCR in any GCP service. Therefore, this is the only way to gain insight into how much we are querying GCR.